VitalHealthConnect/src/main/java/io/tryvital/vitalhealthconnect/workers/ResourceSyncWorker.kt

```diff
@@ -0,0 +1,177 @@
+        build()
+    }
+
+    companion object {
+        fun fromData(data: Data) = ResourceSyncWorkerInput(
+            resource = VitalResource.valueOf(
+                data.getString("resource") ?: throw IllegalArgumentException("Missing resource")
+            ),
+            environment = Environment.valueOf(
+                data.getString("environment") ?: throw IllegalArgumentException("Missing environment")
+            ),
+            region = Region.valueOf(
+                data.getString("region") ?: throw IllegalArgumentException("Missing region")
+            ),
+            apiKey = data.getString("apiKey") ?: throw IllegalArgumentException("Missing API key"),
+        )
+    }
+}
+
+sealed class ResourceSyncState {
+    object Historical : ResourceSyncState()
+    data class Incremental(val changesToken: String, val lastSync: Date) : ResourceSyncState()
+
+    companion object {
+        val adapterFactory: PolymorphicJsonAdapterFactory<ResourceSyncState>
+            get() = PolymorphicJsonAdapterFactory.of(ResourceSyncState::class.java, "type")
+                .withSubtype(Historical::class.java, "historical")
+                .withSubtype(Incremental::class.java, "incremental")
+    }
+}
+
+class ResourceSyncWorker(appContext: Context, workerParams: WorkerParameters) :
+    CoroutineWorker(appContext, workerParams) {
+
+    private val input: ResourceSyncWorkerInput by lazy {
+        ResourceSyncWorkerInput.fromData(inputData)
+    }
+
+    private val vitalClient: VitalClient by lazy {
+        VitalClient(applicationContext, input.region, input.environment, input.apiKey)
+    }
+    private val sharedPreferences: SharedPreferences get() = vitalClient.sharedPreferences
+    private val healthConnectClientProvider by lazy { HealthConnectClientProvider() }
+
+    private val recordReader: RecordReader by lazy {
+        HealthConnectRecordReader(applicationContext, healthConnectClientProvider)
+    }
+
+    private val recordProcessor: RecordProcessor by lazy {
+        HealthConnectRecordProcessor(
+            recordReader,
+            HealthConnectRecordAggregator(applicationContext, healthConnectClientProvider),
+        )
+    }
+
+    private val recordUploader: RecordUploader by lazy {
+        VitalClientRecordUploader(vitalClient)
+    }
+
+    private val vitalLogger = VitalLogger.getOrCreate()
+
+    /**
+     * Each instance of ResourceSyncWorker is responsible to sync one specific VitalResource type.
+     * Sync state of each VitalResource type is stored and advanced independently.
+     *
+     * This separation enables the SDK to function with only partial permission grants (e.g., due to
+     * resource types not needed by the customers, permissions not having been requested, or user
+     * having revoked the permission).
+     *
+     * ## Generic Backfill process
+     * generic_backfill(data_stage, start, end)
+     * 1. Fetch initial changes token T_0.
+     * 2. Fetch data given `[start, end)`
```

Comment:
`[start, end]`?

Reply:
I though the open interval would be `[start, end[`. But:
![image](https://github.com/tryVital/vital-android/assets/949286/219e6f1c-0763-40bb-9630-447173c09512)